File: VideoEffects/UnitTests.WindowsPhone/MediaCompositionTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.VisualStudio.TestPlatform.UnitTestFramework;
using Lumia.Imaging;
using Lumia.Imaging.Artistic;
using Lumia.Imaging.Transforms;
using System.Threading.Tasks;
using VideoEffects;
using Windows.Media.Editing;
using Windows.Media.MediaProperties;
using Windows.Storage;
using Windows.Foundation.Collections;
using Windows.Media.Transcoding;
using Windows.Storage.Streams;
using Windows.Media.Core;
using Windows.Foundation;

namespace UnitTests
{
    [TestClass]
    public class MediaCompositionTests
    {
        // DataTestMethod/DataRow do not seem to work in Phone tests yet, so fake it
        [TestMethod]
        public async Task CS_WP_MC_Basic_Lumia() { await CS_WP_MC_Basic(EffectType.Lumia); }
        [TestMethod]
        public void CS_WP_MC_Basic_ShaderNv12() { Assert.Inconclusive("Does not work in Phone Emulator (no DXGI device manager)"); }
        [TestMethod]
        public void CS_WP_MC_Basic_ShaderBgrx8() { Assert.Inconclusive("Does not work in Phone Emulator (no DXGI device manager)"); }

        async Task CS_WP_MC_Basic(EffectType effectType)
        {
            StorageFile source = await StorageFile.GetFileFromApplicationUriAsync(new Uri("ms-appx:///Input/Car.mp4"));
            StorageFile destination = await KnownFolders.VideosLibrary.CreateFileAsync("CS_WP_MC_Basic_" + effectType + ".mp4", CreationCollisionOption.ReplaceExisting);

            var definition = await Utils.CreateEffectDefinitionAsync(effectType);

            var clip = await MediaClip.CreateFromFileAsync(source);
            clip.VideoEffectDefinitions.Add(definition);

            var composition = new MediaComposition();
            composition.Clips.Add(clip);

            await composition.RenderToFileAsync(destination);
        }

        [DataTestMethod]
        [DataRow(EffectType.Lumia, DisplayName = "Lumia")]
        [DataRow(EffectType.ShaderNv12, DisplayName = "ShaderNv12")]
        [DataRow(EffectType.ShaderBgrx8, DisplayName = "ShaderBgrx8")]
        public async Task CS_WP_MC_PreviewTranscode(EffectType effectType)
        {
            StorageFile source = await StorageFile.GetFileFromApplicationUriAsync(new Uri("ms-appx:///Input/Car.mp4"));
            StorageFile destination = await KnownFolders.VideosLibrary.CreateFileAsync("CS_WP_MC_PreviewTranscode_" + effectType + ".mp4", CreationCollisionOption.ReplaceExisting);

            var definition = await Utils.CreateEffectDefinitionAsync(effectType);

            var clip = await MediaClip.CreateFromFileAsync(source);
            clip.VideoEffectDefinitions.Add(definition);

            var composition = new MediaComposition();
            composition.Clips.Add(clip);

            MediaStreamSource sourceStreamSource = composition.GenerateMediaStreamSource();
            using (IRandomAccessStream destinationStream = await destination.OpenAsync(FileAccessMode.ReadWrite))
            {
                var transcoder = new MediaTranscoder();
                var transcode = await transcoder.PrepareMediaStreamSourceTranscodeAsync(sourceStreamSource, destinationStream, MediaEncodingProfile.CreateMp4(VideoEncodingQuality.Qvga));
                await transcode.TranscodeAsync();
            }
        }

        [TestMethod]
        public async Task CS_WP_MC_LumiaCropSquare()
        {
            StorageFile source = await StorageFile.GetFileFromApplicationUriAsync(new Uri("ms-appx:///Input/Car.mp4"));
            StorageFile destination = await KnownFolders.VideosLibrary.CreateFileAsync("CS_W_MT_CropSquare.mp4", CreationCollisionOption.ReplaceExisting);

            // Select the largest centered square area in the input video
            var encodingProfile = await MediaEncodingProfile.CreateFromFileAsync(source);
            uint inputWidth = encodingProfile.Video.Width;
            uint inputHeight = encodingProfile.Video.Height;
            uint outputLength = Math.Min(inputWidth, inputHeight);
            Rect cropArea = new Rect(
                (float)((inputWidth - outputLength) / 2),
                (float)((inputHeight - outputLength) / 2),
                (float)outputLength,
                (float)outputLength
                );
            encodingProfile.Video.Width = outputLength;
            encodingProfile.Video.Height = outputLength;

            var definition = new LumiaEffectDefinition(new FilterChainFactory(() =>
            {
                var filters = new List<IFilter>();
                filters.Add(new CropFilter(cropArea));
                return filters;
            }));
            definition.InputWidth = inputWidth;
            definition.InputHeight = inputHeight;
            definition.OutputWidth = outputLength;
            definition.OutputHeight = outputLength;

            var clip = await MediaClip.CreateFromFileAsync(source);
            clip.VideoEffectDefinitions.Add(definition);

            var composition = new MediaComposition();
            composition.Clips.Add(clip);

            MediaStreamSource sourceStreamSource = composition.GenerateMediaStreamSource();
            using (IRandomAccessStream destinationStream = await destination.OpenAsync(FileAccessMode.ReadWrite))
            {
                var transcoder = new MediaTranscoder();
                var transcode = await transcoder.PrepareMediaStreamSourceTranscodeAsync(sourceStreamSource, destinationStream, MediaEncodingProfile.CreateMp4(VideoEncodingQuality.Qvga));
                await transcode.TranscodeAsync();
            }
        }
    }
}

File: VideoEffects/UnitTestsCx.WindowsPhone/MediaCompositionTests.cpp
#include "pch.h"

using namespace concurrency;
using namespace Microsoft::VisualStudio::CppUnitTestFramework;
using namespace Lumia::Imaging;
using namespace Lumia::Imaging::Artistic;
using namespace Lumia::Imaging::Transforms;
using namespace Platform;
using namespace Platform::Collections;
using namespace VideoEffects;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::Media;
using namespace Windows::Media::Core;
using namespace Windows::Media::Editing;
using namespace Windows::Media::MediaProperties;
using namespace Windows::Media::Transcoding;
using namespace Windows::Storage;
using namespace Windows::Storage::Streams;

TEST_CLASS(MediaCompositionTests)
{
public:

    TEST_METHOD(CX_WP_MC_Basic)
    {
        StorageFile^ source = Await(StorageFile::GetFileFromApplicationUriAsync(ref new Uri(L"ms-appx:///Car.mp4")));
        StorageFile^ destination = Await(KnownFolders::VideosLibrary->CreateFileAsync(L"CX_WP_MC_Basic.mp4", CreationCollisionOption::ReplaceExisting));

        auto definition = ref new LumiaEffectDefinition(ref new FilterChainFactory([]()
        {
            auto filters = ref new Vector<IFilter^>();
            filters->Append(ref new AntiqueFilter());
            filters->Append(ref new FlipFilter(FlipMode::Horizontal));
            return filters;
        }));

        auto clip = Await(MediaClip::CreateFromFileAsync(source));
        clip->VideoEffectDefinitions->Append(definition);

        auto composition = ref new MediaComposition();
        composition->Clips->Append(clip);

        Await(composition->RenderToFileAsync(destination));
    }

    TEST_METHOD(CX_WP_MC_PreviewTranscode)
    {
        StorageFile^ source = Await(StorageFile::GetFileFromApplicationUriAsync(ref new Uri(L"ms-appx:///Car.mp4")));
        StorageFile^ destination = Await(KnownFolders::VideosLibrary->CreateFileAsync(L"CX_WP_MC_PreviewTranscode.mp4", CreationCollisionOption::ReplaceExisting));

        auto definition = ref new LumiaEffectDefinition(ref new FilterChainFactory([]()
        {
            auto filters = ref new Vector<IFilter^>();
            filters->Append(ref new AntiqueFilter());
            filters->Append(ref new FlipFilter(FlipMode::Horizontal));
            return filters;
        }));

        auto clip = Await(MediaClip::CreateFromFileAsync(source));
        clip->VideoEffectDefinitions->Append(definition);

        auto composition = ref new MediaComposition();
        composition->Clips->Append(clip);

        MediaStreamSource^ sourceStreamSource = composition->GenerateMediaStreamSource();
        IRandomAccessStream^ destinationStream = Await(destination->OpenAsync(FileAccessMode::ReadWrite));

        auto transcoder = ref new MediaTranscoder();

        PrepareTranscodeResult^ transcode = Await(transcoder->PrepareMediaStreamSourceTranscodeAsync(
            sourceStreamSource, 
            destinationStream, 
            MediaEncodingProfile::CreateMp4(VideoEncodingQuality::Qvga)
            ));
        Await(transcode->TranscodeAsync());
    }

    TEST_METHOD(CX_WP_MC_Analysis)
    {
        StorageFile^ source = Await(StorageFile::GetFileFromApplicationUriAsync(ref new Uri(L"ms-appx:///Car.mp4")));
        StorageFile^ destination = Await(KnownFolders::VideosLibrary->CreateFileAsync(L"CX_WP_MC_Analysis.mp4", CreationCollisionOption::ReplaceExisting));

        unsigned int callbackCount = 0;
        auto definition = ref new LumiaAnalyzerDefinition(
            ColorMode::Yuv420Sp,
            320,
            ref new BitmapVideoAnalyzer([&callbackCount](Bitmap^ bitmap, TimeSpan /*time*/)
        {
            Assert::IsNotNull(bitmap);
            Assert::AreEqual((int)ColorMode::Yuv420Sp, (int)bitmap->ColorMode);
            Assert::AreEqual(320, (int)bitmap->Dimensions.Width);
            Assert::AreEqual(240, (int)bitmap->Dimensions.Height);
            Assert::AreEqual(2, (int)bitmap->Buffers->Length);
            InterlockedIncrement(&callbackCount);
        }));

        auto clip = Await(MediaClip::CreateFromFileAsync(source));
        clip->VideoEffectDefinitions->Append(definition);

        auto composition = ref new MediaComposition();
        composition->Clips->Append(clip);

        Await(composition->RenderToFileAsync(destination));

        Log() << L"Callback count: " << callbackCount;

        Assert::IsTrue(callbackCount > 0);
    }

};
